lib/path-design.md

```diff
@@ -0,0 +1,373 @@
+
+Laws:
+- Associativity:
+  `relative.join [ x (join [ y z ]) ] == relative.join [ (join [ x y ]) z ]`
+- Identity:
+  `relative.join [] == "."`
+  `relative.join [p "."] == normalise p`
+  `relative.join ["." p] == normalise p`
+- The result is normalised:
+  `relative.join ps == normalise (relative.join ps)`
+- Joining a single path is that path itself, but normalised:
+  `relative.join [ p ] == normalise p`
+
+### `relative.normalise`
+
+```haskell
+relative.normalise :: String -> String
+```
+
+Normalise relative paths.
+
+- Limit repeating `/` to a single one (does not change a [POSIX Pathname](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_271))
+- Remove redundant `.` components (does not change the result of [POSIX Pathname Resolution](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13))
+- Error on empty strings (not allowed as a [POSIX Filename](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_170))
+- Remove trailing `/` and `/.` (See [justification](#trailing-slashes))
+- Error on `..` path components (See [justification](#parents))
+- Remove leading `./` (See [justification](#leading-dots))
+
+Examples:
+```
+# limit repeating `/` to a single one
+relative.normalise "foo//bar" == "foo/bar"
+
+# remove redundant `.` components
+relative.normalise "foo/./bar" == "foo/bar"
+
+# remove leading `./`
+# TODO: bikeshedding. there is a use to this.
+relative.normalise "./foo/bar" == "foo/bar"
+
+# remove trailing `/`
+relative.normalise "foo/bar/" == "foo/bar"
+
+# remove trailing `/.`
+relative.normalise "foo/bar/." == "foo/bar"
+
+# error on `..` path components
+relative.normalise "foo/../bar" == <error>
+
+# error on empty string
+relative.normalise "" == <error>
+
+# error on absolute path
+relative.normalise "/foo" == <error>
+```
+
+Laws:
+- Idempotency:
+
+  `relative.normalise (relative.normalise p) == relative.normalise p`
+
+- Doesn't change the file system object pointed to:
+
+  `$(stat ${p}) == $(stat ${relative.normalise p})`
+
+  TODO: Is this the same as the law below?
+
+- Uniqueness: If the normalisation of two paths is different then they point to different paths:
+
+  `normalise p != normalise q => $(stat ${p}) != $(stat ${q})`
+
+  Note: This law only holds if trailing slashes are not persisted, see the [trailing slashes decision](#trailing-slashes)
+
+### `difference`
+
+```haskell
+difference :: AttrsOf Path -> { commonPrefix :: Path; suffix :: AttrsOf String; }
+```
+
+Take the difference between multiple paths, returning the common prefix between them and the respective suffices.
+
+Examples:
+```nix
+difference { path = /foo/bar } = { commonPrefix = /foo/bar; suffix = { path = "."; }; }
+difference { left = /foo/bar; right = /foo/baz; } = { commonPrefix = /foo; suffix = { left = "bar"; right = "baz"; }; }
+difference { left = /foo; right = /foo/bar; } = { commonPrefix = /foo; suffix = { left = "."; right = "bar"; }; }
```

Comment:
```suggestion
difference { a = /foo; b = /foo/bar; c = /foo/baz } = { commonPrefix = /foo; suffix = { a = "."; b = "bar"; c = "baz"; }; }
```

Illustrate that the names are up to the caller, for those who don't normally read types.